core/src/main/scala/chisel3/StrongEnum.scala

```diff
@@ -288,15 +291,37 @@ abstract class EnumFactory {
     } else if (n.getWidth > this.getWidth) {
       throwException(s"The UInt being cast to $enumTypeName is wider than $enumTypeName's width ($getWidth)")
     } else {
-      Builder.warning(s"Casting non-literal UInt to $enumTypeName. You can check that its value is legal by calling isValid")
-
+      // Check if its possible for `n` to hit an illegal state
+      val canHitIllegalState = warn && !this.isTotal
+      if (canHitIllegalState) {
```

Comment:
Nit: consider inlining since this has one use.

Reply:
It was longer before I factored `isTotal` out into a method which is why I named it originally, I'll inline it.

Reply:
Done.